# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1536870258 14400
#      Thu Sep 13 16:24:18 2018 -0400
# Node ID 22ee68edcf3a23cd3cfd43c87d063206248f482e
# Parent  369e43cda3c7e616c02441479e1df8e9b5c532f3
require dynamic loading capabilities to build Octave

* configure.ac: Eliminate --disable-dl configure option.
(ENABLE_DYNAMIC_LINKING): Eliminate variable, corresponding automake
conditiona, and all uses.  Require dlopen, shl_load, loadlibrary, or
dyld api to build Octave.

* install.txi: Don't document --enable-dl.

* mex_demo.c, oct_demo.cc: Don't mention ENABLE_DYNAMIC_LINKING in
comments.

* toplev.cc (F__octave_config_info__): Remove ENABLE_DYNAMIC_LINKING
and dld from from structs.

* libinterp/module.mk, mk-builtins.pl, config-module.awk:
Always assume dynamic linking is used.

diff --git a/configure.ac b/configure.ac
--- a/configure.ac
+++ b/configure.ac
@@ -643,31 +643,16 @@ AC_SUBST(STATIC_LIBS)
 
 XTRA_EXTERNAL_SH_LDFLAGS=
 if test $have_msvc = yes; then
   FLIBS="$FLIBS -lkernel32"
   XTRA_EXTERNAL_SH_LDFLAGS="-Wl,external/external.def"
 fi
 AC_SUBST(XTRA_EXTERNAL_SH_LDFLAGS)
 
-### Enable dynamic linking.
-### --enable-shared implies this, thus --enable-dl is only needed if you are
-### only building static libraries and want to try dynamic linking too
-### (works on some systems, for example, OS X and Windows).
-
-AC_ARG_ENABLE([dl],
-  [AS_HELP_STRING([--disable-dl],
-    [disable loading of dynamically linked modules])],
-  [case $enableval in
-     yes) ENABLE_DYNAMIC_LINKING=yes ;;
-     no) ENABLE_DYNAMIC_LINKING=no ;;
-     *) AC_MSG_ERROR([bad value $enableval for --enable-dl]) ;;
-   esac],
-  [ENABLE_DYNAMIC_LINKING=no])
-
 if test $STATIC_LIBS = no && test $SHARED_LIBS = no; then
   AC_MSG_ERROR([You can't disable building both static AND shared libraries!])
 fi
 
 
 ### More configure argument checking related to linking
 
 AC_ARG_ENABLE([no-undefined],
@@ -864,79 +849,59 @@ dnl code generated by the dlopen option 
 LD_CXX="${CXX}"
 RDYNAMIC_FLAG=
 DL_API_MSG=""
 dlopen_api=no
 shl_load_api=no
 loadlibrary_api=no
 dyld_api=no
 
-if test $SHARED_LIBS = yes || test $ENABLE_DYNAMIC_LINKING = yes; then
-
-  case $lt_cv_dlopen in
-    dlopen)
-      dlopen_api=yes
-      DL_API_MSG="(dlopen)"
-      AC_DEFINE(HAVE_DLOPEN_API, 1,
-        [Define to 1 if system has dlopen, dlsym, dlerror, and dlclose for dynamic linking.])
-      OCTAVE_CXX_FLAG([-rdynamic], [RDYNAMIC_FLAG=-rdynamic])
-    ;;
-    shl_load)
-      shl_load_api=yes
-      DL_API_MSG="(shl_load)"
-      AC_DEFINE(HAVE_SHL_LOAD_API, 1,
-        [Define to 1 if system has shl_load and shl_findsym for dynamic linking.])
-    ;;
-    LoadLibrary)
-      loadlibrary_api=yes
-      DL_API_MSG="(LoadLibrary)"
-      AC_DEFINE(HAVE_LOADLIBRARY_API, 1,
-        [Define to 1 if system has LoadLibrary for dynamic linking.])
-    ;;
-    dyld)
-      dyld_api=yes
-      DL_API_MSG="(dyld)"
-      AC_DEFINE(HAVE_DYLD_API, 1,
-        [Define to 1 if system has dyld for dynamic linking.])
-    ;;
-  esac
-
-  DL_LIBS="$lt_cv_dlopen_libs"
-  AC_SUBST(DL_LIBS)
-
-  ## Disable dynamic linking if capability is not present.
-  if test $dlopen_api = yes \
-      || test $shl_load_api = yes \
-      || test $loadlibrary_api = yes \
-      || test $dyld_api = yes; then
-    ## some form of dynamic linking present
-    ENABLE_DYNAMIC_LINKING=yes
-  else
-    ENABLE_DYNAMIC_LINKING=no
-  fi
-fi
-
-if test $ENABLE_DYNAMIC_LINKING = yes; then
-  AC_DEFINE(ENABLE_DYNAMIC_LINKING, 1, [Define to 1 if using dynamic linking.])
-fi
-
-AM_CONDITIONAL([AMCOND_ENABLE_DYNAMIC_LINKING],
-  [test $ENABLE_DYNAMIC_LINKING = yes])
+case $lt_cv_dlopen in
+  dlopen)
+    dlopen_api=yes
+    DL_API_MSG="dlopen"
+    AC_DEFINE(HAVE_DLOPEN_API, 1,
+      [Define to 1 if system has dlopen, dlsym, dlerror, and dlclose for dynamic linking.])
+    OCTAVE_CXX_FLAG([-rdynamic], [RDYNAMIC_FLAG=-rdynamic])
+  ;;
+  shl_load)
+    shl_load_api=yes
+    DL_API_MSG="shl_load"
+    AC_DEFINE(HAVE_SHL_LOAD_API, 1,
+      [Define to 1 if system has shl_load and shl_findsym for dynamic linking.])
+  ;;
+  LoadLibrary)
+    loadlibrary_api=yes
+    DL_API_MSG="LoadLibrary"
+    AC_DEFINE(HAVE_LOADLIBRARY_API, 1,
+      [Define to 1 if system has LoadLibrary for dynamic linking.])
+  ;;
+  dyld)
+    dyld_api=yes
+    DL_API_MSG="dyld"
+    AC_DEFINE(HAVE_DYLD_API, 1,
+      [Define to 1 if system has dyld for dynamic linking.])
+  ;;
+  *)
+    AC_MSG_ERROR([Octave requires some way to perform dynamic linking.])
+  ;;
+esac
+
+DL_LIBS="$lt_cv_dlopen_libs"
 
 if test $SHARED_LIBS = yes; then
   LIBOCTINTERP="-loctinterp"
   LIBOCTAVE="-loctave"
 else
   LIBOCTINTERP="${top_builddir}/libinterp/liboctinterp.a"
   LIBOCTAVE="${top_builddir}/liboctave/liboctave.a"
 fi
 
 AC_SUBST(LD_CXX)
 AC_SUBST(RDYNAMIC_FLAG)
-AC_SUBST(ENABLE_DYNAMIC_LINKING)
 AC_SUBST(LIBOCTINTERP)
 AC_SUBST(LIBOCTAVE)
 
 if test "$cross_compiling" = yes && test -n "$ac_tool_prefix"; then
   CROSS_TOOL_PREFIX="$ac_tool_prefix"
   MKOCTFILE_AR=`echo "$AR" | $SED "s,$CROSS_TOOL_PREFIX,,"`
   MKOCTFILE_CC=`echo "$CC" | $SED "s,$CROSS_TOOL_PREFIX,,"`
   MKOCTFILE_CXX=`echo "$CXX" | $SED "s,$CROSS_TOOL_PREFIX,,"`
@@ -2771,23 +2736,17 @@ LIBOCTAVE_LINK_DEPS="$BASE_LIBOCTAVE_LIN
 
 MKOCTFILE_LIBOCTAVE_LINK_DEPS="$BASE_LIBOCTAVE_LINK_DEPS $MKOCTFILE_GNULIB_LINK_DEPS"
 
 LIBOCTAVE_LINK_OPTS="$CURL_LDFLAGS $SPARSE_XLDFLAGS $ARPACK_LDFLAGS $QRUPDATE_LDFLAGS $FFTW_XLDFLAGS $PCRE_LDFLAGS"
 
 AC_SUBST(LIBOCTAVE_LINK_DEPS)
 AC_SUBST(LIBOCTAVE_LINK_OPTS)
 
-if test $ENABLE_DYNAMIC_LINKING = yes; then
-  LIBOCTINTERP_LINK_DEPS=""
-else
-  LIBOCTINTERP_LINK_DEPS="$DLDFCN_LIBS"
-fi
-
-LIBOCTINTERP_LINK_DEPS="$LIBOCTINTERP_LINK_DEPS $FT2_LIBS $HDF5_LIBS $MAGICK_LIBS $Z_LIBS $FFTW_XLIBS $OPENGL_LIBS $FONTCONFIG_LIBS $FREETYPE_LIBS $X11_LIBS $CARBON_LIBS $GL2PS_LIBS $LLVM_LIBS $JAVA_LIBS $LAPACK_LIBS"
+LIBOCTINTERP_LINK_DEPS="$FT2_LIBS $HDF5_LIBS $MAGICK_LIBS $Z_LIBS $FFTW_XLIBS $OPENGL_LIBS $FONTCONFIG_LIBS $FREETYPE_LIBS $X11_LIBS $CARBON_LIBS $GL2PS_LIBS $LLVM_LIBS $JAVA_LIBS $LAPACK_LIBS"
 
 LIBOCTINTERP_LINK_OPTS="$FT2_LDFLAGS $HDF5_LDFLAGS $MAGICK_LDFLAGS $Z_LDFLAGS $FFTW_XLDFLAGS $LLVM_LDFLAGS"
 
 OCTAVE_LINK_DEPS=""
 OCTAVE_LINK_OPTS=""
 
 OCT_LINK_DEPS=""
 OCT_LINK_OPTS="$LDFLAGS"
@@ -2825,23 +2784,17 @@ AC_SUBST(MKOCTFILE_OCT_LINK_DEPS)
 
 LIBOCTGUI_LINK_DEPS=""
 LIBOCTGUI_LINK_OPTS=""
 
 OCTAVE_GUI_LINK_DEPS=""
 OCTAVE_GUI_LINK_OPTS=""
 
 if test $build_qt_gui = yes; then
-  if test $ENABLE_DYNAMIC_LINKING = yes; then
-    LIBOCTGUI_LINK_DEPS=""
-  else
-    LIBOCTGUI_LINK_DEPS="$DLDFCN_LIBS"
-  fi
-
-  LIBOCTGUI_LINK_DEPS="$LIBOCTGUI_LINK_DEPS $QT_LIBS $OPENGL_LIBS"
+  LIBOCTGUI_LINK_DEPS="$QT_LIBS $OPENGL_LIBS"
   LIBOCTGUI_LINK_OPTS="$QT_LDFLAGS"
 
   if test $link_all_deps = yes || test -n "$QT_LDFLAGS"; then
     LIBOCTGUI_LINK_DEPS="$LIBOCTGUI_LINK_DEPS $LIBOCTINTERP_LINK_DEPS"
     LIBOCTGUI_LINK_OPTS="$LIBOCTGUI_LINK_OPTS $LIBOCTINTERP_LINK_OPTS"
 
     OCTAVE_GUI_LINK_DEPS="$OCTAVE_GUI_LINK_DEPS $LIBOCTGUI_LINK_DEPS"
     OCTAVE_GUI_LINK_OPTS="$OCTAVE_GUI_LINK_OPTS $LIBOCTGUI_LINK_OPTS"
@@ -3106,39 +3059,30 @@ Octave is now configured for $canonical_
   Default pager:                 $DEFAULT_PAGER
   gnuplot:                       $GNUPLOT
 
   Build Octave Qt GUI:                  $BUILD_QT_SUMMARY_MSG
   JIT compiler for loops:               $ENABLE_JIT
   Build Java interface:                 $build_java
   Build static libraries:               $STATIC_LIBS
   Build shared libraries:               $SHARED_LIBS
-  Dynamic Linking:                      $ENABLE_DYNAMIC_LINKING $DL_API_MSG
+  Dynamic Linking API:                  $DL_API_MSG
   Include support for GNU readline:     $USE_READLINE
   64-bit array dims and indexing:       $ENABLE_64
   64-bit BLAS array dims and indexing:  $HAVE_64_BIT_BLAS
   OpenMP SMP multithreading:            $ENABLE_OPENMP
   Truncate intermediate FP results:     $ENABLE_FLOAT_TRUNCATE
   Build cross tools:                    $cross_tools
   Build docs:                           $ENABLE_DOCS
 ])
 
 warn_msg_printed=false
 
 OCTAVE_CONFIGURE_WARNING_SUMMARY
 
-if test $ENABLE_DYNAMIC_LINKING = yes; then
-  if test $SHARED_LIBS = no; then
-    AC_MSG_WARN([])
-    AC_MSG_WARN([You used --enable-dl but not --enable-shared.])
-    AC_MSG_WARN([Are you sure that is what you want to do?])
-    warn_msg_printed=true
-  fi
-fi
-
 if test $opengl_graphics = no; then
   AC_MSG_WARN([])
   AC_MSG_WARN([The libraries needed for OpenGL graphics were not found.])
   AC_MSG_WARN([Creating plots is still possible if gnuplot is installed.])
   warn_msg_printed=true
 fi
 
 if test $ENABLE_ATOMIC_REFCOUNT = no; then
diff --git a/doc/interpreter/install.txi b/doc/interpreter/install.txi
--- a/doc/interpreter/install.txi
+++ b/doc/interpreter/install.txi
@@ -414,24 +414,16 @@ Create shared libraries (this is the def
 use the dynamic loading features, you will probably want to use this
 option.  It will make your @file{.oct} files much smaller and on some
 systems it may be necessary to build shared libraries in order to use
 dynamically linked functions.
 
 You may also want to build a shared version of @code{libstdc++}, if your
 system doesn't already have one.
 
-@item --enable-dl
-Use @code{dlopen} and friends to make Octave capable of dynamically
-linking externally compiled functions (this is the default if
-@option{--enable-shared} is specified).  This option only works on
-systems that actually have these functions.  If you plan on using this
-feature, you should probably also use @option{--enable-shared} to reduce
-the size of your @file{.oct} files.
-
 @item --with-blas=<lib>
 By default, configure looks for the best @sc{blas} matrix libraries on
 your system, including optimized implementations such as the free ATLAS
 3.0, as well as vendor-tuned libraries.  (The use of an optimized
 @sc{blas} will generally result in several-times faster matrix
 operations.)  Use this option to specify a particular @sc{blas} library
 that Octave should use.
 
diff --git a/examples/code/mex_demo.c b/examples/code/mex_demo.c
--- a/examples/code/mex_demo.c
+++ b/examples/code/mex_demo.c
@@ -1,18 +1,10 @@
 // mex_demo.c -- example of a dynamically linked function for Octave.
 
-// To use this file, your version of Octave must support dynamic
-// linking.  To find out if it does, type the command
-//
-//   __octave_config_info__ ("ENABLE_DYNAMIC_LINKING")
-//
-// at the Octave prompt.  Support for dynamic linking is included if
-// this expression is true.
-//
 // To compile this file, type the command
 //
 //   mkoctfile --mex mex_demo.c
 //
 // from within Octave or from the shell prompt.  This will create a file
 // called mex_demo.mex that can be loaded by Octave.  To test the mex_demo.mex
 // file, start Octave and type the command
 //
diff --git a/examples/code/oct_demo.cc b/examples/code/oct_demo.cc
--- a/examples/code/oct_demo.cc
+++ b/examples/code/oct_demo.cc
@@ -1,18 +1,10 @@
 // oct_demo.cc -- example of a dynamically linked function for Octave.
 
-// To use this file, your version of Octave must support dynamic
-// linking.  To find out if it does, type the command
-//
-//   __octave_config_info__ ("ENABLE_DYNAMIC_LINKING")
-//
-// at the Octave prompt.  Support for dynamic linking is included if
-// this expression is true.
-//
 // To compile this file, type the command
 //
 //   mkoctfile oct_demo.cc
 //
 // from within Octave or from the shell prompt.  This will create a file
 // called oct_demo.oct that can be loaded by Octave.  To test the
 // oct_demo.oct file, start Octave and type the command
 //
diff --git a/libinterp/corefcn/toplev.cc b/libinterp/corefcn/toplev.cc
--- a/libinterp/corefcn/toplev.cc
+++ b/libinterp/corefcn/toplev.cc
@@ -342,22 +342,16 @@ Return a structure containing configurat
 Octave.
 
 If @var{option} is a string, return the configuration information for the
 specified option.
 
 @seealso{computer}
 @end deftypefn */)
 {
-#if defined (ENABLE_DYNAMIC_LINKING)
-  bool octave_supports_dynamic_linking = true;
-#else
-  bool octave_supports_dynamic_linking = false;
-#endif
-
   static octave_scalar_map config;
   static octave_scalar_map build_env;
   static octave_scalar_map build_features;
 
   static bool initialized = false;
 
   if (! initialized)
     {
@@ -377,22 +371,16 @@ specified option.
 #endif
 
 #if defined (ENABLE_DOCS)
            { "ENABLE_DOCS", true },
 #else
            { "ENABLE_DOCS", false },
 #endif
 
-#if defined (ENABLE_DYNAMIC_LINKING)
-           { "ENABLE_DYNAMIC_LINKING", true },
-#else
-           { "ENABLE_DYNAMIC_LINKING", false },
-#endif
-
 #if defined (OCTAVE_ENABLE_FLOAT_TRUNCATE)
            { "ENABLE_FLOAT_TRUNCATE", true },
 #else
            { "ENABLE_FLOAT_TRUNCATE", false },
 #endif
 
 #if defined (ENABLE_JIT)
            { "ENABLE_JIT", true },
@@ -603,18 +591,16 @@ specified option.
 #if defined (OCTAVE_USE_OS_X_API)
       mac_system = true;
 #endif
 
       config.assign ("unix", octave_value (unix_system));
       config.assign ("mac", octave_value (mac_system));
       config.assign ("windows", octave_value (windows_system));
 
-      config.assign ("dld", octave_value (octave_supports_dynamic_linking));
-
       octave::mach_info::float_format ff = octave::mach_info::native_float_format ();
       config.assign ("float_format",
                      octave_value (octave::mach_info::float_format_as_string (ff)));
 
       config.assign ("words_big_endian",
                      octave_value (octave::mach_info::words_big_endian ()));
 
       config.assign ("words_little_endian",
diff --git a/libinterp/dldfcn/config-module.awk b/libinterp/dldfcn/config-module.awk
--- a/libinterp/dldfcn/config-module.awk
+++ b/libinterp/dldfcn/config-module.awk
@@ -44,31 +44,23 @@ BEGIN {
       sep = "\n";
     printf ("  %%reldir%%/%s%s", files[i], sep);
   }
   print "";
 
   sep = " \\\n";
   print "DLDFCN_LIBS = $(DLDFCN_SRC:.cc=.la)";
   print "";
-  print "if AMCOND_ENABLE_DYNAMIC_LINKING";
-  print "";
   print "octlib_LTLIBRARIES += $(DLDFCN_LIBS)";
   print "";
   print "## Use stamp files to avoid problems with checking timestamps";
   print "## of symbolic links";
   print "";
   print "%.oct : %.la"
   print "\t$(AM_V_GEN)$(INSTALL_PROGRAM) %reldir%/.libs/$(shell $(SED) -n -e \"s/dlname='\\([^']*\\)'/\\1/p\" < $<) $@"
-  print ""
-  print "else";
-  print "";
-  print "noinst_LTLIBRARIES += $(DLDFCN_LIBS)";
-  print "";
-  print "endif";
 
   for (i = 1; i <= nfiles; i++) {
     basename = files[i];
     sub (/\.cc$/, "", basename);
     print "";
     printf ("%%canon_reldir%%_%s_la_SOURCES = %%reldir%%/%s\n",
             basename, files[i]);
     if (cppflags[i])
diff --git a/libinterp/mk-builtins.pl b/libinterp/mk-builtins.pl
--- a/libinterp/mk-builtins.pl
+++ b/libinterp/mk-builtins.pl
@@ -13,17 +13,16 @@
 # WITHOUT ANY WARRANTY; without even the implied warranty of
 # MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 # GNU General Public License for more details.
 #
 # You should have received a copy of the GNU General Public License
 # along with Octave; see the file COPYING.  If not, see
 # <https://www.gnu.org/licenses/>.
 
-$defun_dld_are_built_in = 0;
 $make_header = 0;
 $make_source = 0;
 
 $have_options = 1;
 
 while ($have_options)
 {
   $opt = shift (@ARGV);
@@ -35,33 +34,26 @@ while ($have_options)
       $make_header = 1;
     }
   elsif ($opt eq "--source")
     {
       die "mk-builtins.pl: only one of --header or --source may be specified" if ($make_header);
 
       $make_source = 1;
     }
-  elsif ($opt eq "--disable-dl")
-    {
-      ## If DLD functions are disabled, then DEFUN_DLD functions are
-      ## built-in instead of being dynamically loaded so they will also
-      ## need to be installed.
-      $defun_dld_are_built_in= 1;
-    }
   else
     {
       $srcdir = "$opt";
       $opt = shift (@ARGV);
       die "mk-builtins.pl: '--' must separate SRCDIR from other file names" if ($opt ne "--");
       $have_options = 0;
     }
 }
 
-die "usage: mk-builtins.pl --header|--source [--disable-dl] SRCDIR -- f1 f2 ..." if (! @ARGV);
+die "usage: mk-builtins.pl --header|--source SRCDIR -- f1 f2 ..." if (! @ARGV);
 
 die "mk-builtins.pl: one of --header or --source must be specified" if (! $make_header && ! $make_source);
 
 if ($make_header)
 {
   print "// DO NOT EDIT!  Generated automatically by mk-builtins.pl.
 
 #if ! defined (octave_builtin_defun_decls_h)
@@ -98,29 +90,16 @@ namespace octave
         $name = "F$2";
         $is_method = ($1 eq "METHOD" || $1 eq "CONSTMETHOD");
       }
       elsif (/^[ \t]*DEF(METHOD|UN)X[ \t]*\( *"[^"]*" *, *([^ ,]*).*$/)
       {
         $name = $2;
         $is_method = ($1 eq "METHOD");
       }
-      elsif ($defun_dld_are_built_in)
-      {
-        if (/^[ \t]*DEF(METHOD|UN)_DLD[ \t]*\( *([^ ,]*).*$/)
-        {
-          $name = "F$2";
-          $is_method = ($1 eq "METHOD");
-        }
-        elsif (/^[ \t]*DEF(METHOD|UN)X_DLD[ \t]*\( *"[^"]*" *, *([^ ,]*).*$/)
-        {
-          $name = "$2";
-          $is_method = ($1 eq "METHOD");
-        }
-      }
 
       if ($name)
       {
         if ($is_method)
         {
           print "extern OCTINTERP_API octave_value_list
 $name (octave::interpreter&, const octave_value_list& = octave_value_list (), int = 0);
 ";
@@ -214,31 +193,16 @@ namespace octave
         $const_param = ", true";
       }
       elsif ($line =~ /^ *DEFALIAS *\( *([^ ,]*) *, *([^ )]*) *\).*$/)
       {
         $type = "alias";
         $alias = "$1";
         $name = "$2";
       }
-      elsif ($defun_dld_are_built_in)
-      {
-        if ($line =~ /^ *DEF(METHOD|UN)_DLD *\( *([^ ,]*) *,.*$/)
-        {
-          $type = "fun";
-          $fname = "F$2";
-          $name = "$2";
-        }
-        elsif ($line =~ /^ *DEF(METHOD|UN)X_DLD *\( *"([^"]*)" *, *([^ ,]*) *,.*$/)
-        {
-          $type = "fun";
-          $fname = "$3";
-          $name = "$2";
-        }
-      }
 
       if ($type eq "fun")
       {
         if (($line = <$fh>) =~ /^ *classes:/)
         {
           $line =~ s/\s*classes:\s*//;
           $line =~ s/\s*$//;
           @classes = split (/\s+/, $line);
diff --git a/libinterp/module.mk b/libinterp/module.mk
--- a/libinterp/module.mk
+++ b/libinterp/module.mk
@@ -108,25 +108,19 @@ DIST_SRC += \
 
 include %reldir%/parse-tree/module.mk
 include %reldir%/octave-value/module.mk
 include %reldir%/operators/module.mk
 include %reldir%/template-inst/module.mk
 include %reldir%/corefcn/module.mk
 include %reldir%/dldfcn/module.mk
 
-if AMCOND_ENABLE_DYNAMIC_LINKING
-  OCT_FILES = $(DLDFCN_LIBS:.la=.oct)
-  DLD_LIBOCTINTERP_LIBADD = $(OCT_LINK_DEPS)
-  LIBINTERP_DLDFCN_LIBADD =
-else
-  OCT_FILES =
-  DLD_LIBOCTINTERP_LIBADD =
-  LIBINTERP_DLDFCN_LIBADD = $(DLDFCN_LIBS)
-endif
+OCT_FILES = $(DLDFCN_LIBS:.la=.oct)
+DLD_LIBOCTINTERP_LIBADD = $(OCT_LINK_DEPS)
+LIBINTERP_DLDFCN_LIBADD =
 
 %canon_reldir%_liboctinterp_la_SOURCES = \
   %reldir%/octave.cc \
   $(LIBINTERP_OPERATORS_SRC) \
   $(TEMPLATE_INST_SRC)
 
 nodist_%canon_reldir%_liboctinterp_la_SOURCES = \
   %reldir%/build-env-features.cc \
@@ -175,21 +169,17 @@ ULT_DIST_SRC := \
 
 LIBINTERP_FOUND_DEFUN_FILES := \
   $(shell $(SHELL) $(srcdir)/build-aux/find-defun-files.sh "$(srcdir)" $(ULT_DIST_SRC))
 
 BUILT_IN_DEFUN_FILES = $(OPT_HANDLERS) $(LIBINTERP_FOUND_DEFUN_FILES)
 
 DLDFCN_DEFUN_FILES = $(DLDFCN_SRC)
 
-if AMCOND_ENABLE_DYNAMIC_LINKING
-  DEFUN_FILES = $(BUILT_IN_DEFUN_FILES)
-else
-  DEFUN_FILES = $(BUILT_IN_DEFUN_FILES) $(DLDFCN_DEFUN_FILES)
-endif
+DEFUN_FILES = $(BUILT_IN_DEFUN_FILES)
 
 LIBINTERP_DEFUN_FILES = $(BUILT_IN_DEFUN_FILES) $(DLDFCN_DEFUN_FILES)
 
 ## FIXME: The following two variables are deprecated and should be removed
 ##        in Octave version 3.12.
 DLL_CDEFS = @OCTINTERP_DLL_DEFS@
 DLL_CXXDEFS = @OCTINTERP_DLL_DEFS@
 
@@ -225,40 +215,34 @@ nobase_libinterptests_DATA = $(LIBINTERP
 %reldir%/build-env-features.cc: config.h %reldir%/mk-build-env-features.sh | %reldir%/$(octave_dirstamp)
 	$(AM_V_GEN)rm -f $@-t && \
 	$(SHELL) $(srcdir)/%reldir%/mk-build-env-features.sh $< > $@-t && \
 	mv $@-t $@
 
 %reldir%/liboctinterp-build-info.cc: %reldir%/liboctinterp-build-info.in.cc HG-ID | %reldir%/$(octave_dirstamp)
 	$(AM_V_GEN)$(build-info-commands)
 
-if AMCOND_ENABLE_DYNAMIC_LINKING
-  mkbuiltins_dld_opt =
-else
-  mkbuiltins_dld_opt = --disable-dl
-endif
+mkbuiltins_dld_opt =
 
 %reldir%/builtins.cc: $(LIBINTERP_DEFUN_FILES) %reldir%/mk-builtins.pl | %reldir%/$(octave_dirstamp)
 	$(AM_V_GEN)rm -f $@-t && \
 	$(PERL) $(srcdir)/%reldir%/mk-builtins.pl --source $(mkbuiltins_dld_opt) "$(srcdir)" -- $(LIBINTERP_DEFUN_FILES) > $@-t && \
 	mv $@-t $@
 
 %reldir%/builtin-defun-decls.h: $(LIBINTERP_DEFUN_FILES) %reldir%/mk-builtins.pl | %reldir%/$(octave_dirstamp)
 	$(AM_V_GEN)rm -f $@-t && \
 	$(PERL) $(srcdir)/%reldir%/mk-builtins.pl --header $(mkbuiltins_dld_opt) "$(srcdir)" -- $(LIBINTERP_DEFUN_FILES) > $@-t && \
 	$(simple_move_if_change_rule)
 
-if AMCOND_ENABLE_DYNAMIC_LINKING
 DLDFCN_PKG_ADD_FILE = %reldir%/dldfcn/PKG_ADD
 
 %reldir%/dldfcn/PKG_ADD: $(DLDFCN_DEFUN_FILES) %reldir%/mk-pkg-add.sh | %reldir%/$(octave_dirstamp)
 	$(AM_V_GEN)rm -f $@-t && \
 	$(SHELL) $(srcdir)/%reldir%/mk-pkg-add.sh "$(srcdir)" $(DLDFCN_DEFUN_FILES) > $@-t && \
 	mv $@-t $@
-endif
 
 DOCSTRING_FILES += %reldir%/DOCSTRINGS
 
 %reldir%/DOCSTRINGS: $(LIBINTERP_DEFUN_FILES) %reldir%/op-kw-docs | %reldir%/$(octave_dirstamp)
 	$(AM_V_GEN)rm -f %reldir%/DOCSTRINGS-t && \
 	( $(PERL) $(srcdir)/%reldir%/mk-doc.pl "$(srcdir)" $(LIBINTERP_DEFUN_FILES); $(SED) -ne '/^\x1d/,$$p' $(srcdir)/%reldir%/op-kw-docs ) > %reldir%/DOCSTRINGS-t && \
 	$(call move_if_change_rule,%reldir%/DOCSTRINGS-t,$@)
 
@@ -268,17 +252,16 @@ OCTAVE_INTERPRETER_TARGETS += \
   $(LIBINTERP_TST_FILES)
 
 DIRSTAMP_FILES += %reldir%/$(octave_dirstamp)
 
 install-data-hook: install-oct install-built-in-docstrings
 
 uninstall-local: uninstall-oct uninstall-built-in-docstrings
 
-if AMCOND_ENABLE_DYNAMIC_LINKING
 install-oct:
 	$(MKDIR_P) $(DESTDIR)$(octfiledir)
 	if [ -n "`cat $(DLDFCN_PKG_ADD_FILE)`" ]; then \
 	  $(INSTALL_DATA) $(DLDFCN_PKG_ADD_FILE) $(DESTDIR)$(octfiledir)/PKG_ADD; \
 	fi
 	cd $(DESTDIR)$(octlibdir) && \
 	for ltlib in $(DLDFCN_LIBS); do \
 	  f=`echo $$ltlib | $(SED) 's,.*/,,'`; \
@@ -289,24 +272,24 @@ install-oct:
 	    echo "error: dlname is empty in $$ltlib!"; \
 	    exit 1; \
 	  fi; \
 	  lnames=`$(SED) -n -e "s/library_names='\([^']*\)'/\1/p" < $$f`; \
 	  if [ -n "$$lnames" ]; then \
 	    rm -f $$f $$lnames $$dl; \
 	  fi \
 	done
+.PHONY: install-oct
 
 uninstall-oct:
 	for f in $(notdir $(OCT_FILES)); do \
 	  rm -f $(DESTDIR)$(octfiledir)/$$f; \
 	done
 	rm -f $(DESTDIR)$(octfiledir)/PKG_ADD
-endif
-.PHONY: install-oct uninstall-oct
+.PHONY: uninstall-oct
 
 install-built-in-docstrings: %reldir%/DOCSTRINGS
 	$(MKDIR_P) $(DESTDIR)$(octetcdir)
 	$(INSTALL_DATA) $< $(DESTDIR)$(octetcdir)/built-in-docstrings
 .PHONY: install-built-in-docstrings
 
 uninstall-built-in-docstrings:
 	rm -f $(DESTDIR)$(octetcdir)/built-in-docstrings
